### PR TITLE
Refactoring node section

### PR DESCRIPTION
**Pull request by borismelnik**
refactor(node): Move node section to new architecture

Closes: #3805

### CODE REVIEW

**Review by saledjenic — CHANGES_REQUESTED** [review]
@borismelnik I added some observations, we may discuss them further if needed. 

- The main thing you were able to build the project and why you haven't seen errors is cause your module has not been created/called ever. So bring it to the app and connect it appropriately.

- One more general suggestion is that setting delegate for a child module, in case of Nim, is maybe easier/cleaner using `AccessInterface` than the concepts, and also that will enable using third, fourth levels of submodules if needed, while with concepts only 2 sub levels is doable.

- Another general thing is about naming interfaces files in `private_interfaces`, I don't have anything against the way you used, but maybe we should stick to the naming pattern already used in other `private_interfaces` folders.

File: src/app/modules/main/node_section/controller.nim [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import ../../../../../app_service/service/settings/service as settings_service
+import ../../../../../app_service/service/node/service as node_service
+
+import eventemitter
+import status/[signals]
+
+export controller_interface
+
+type
+  Controller*[T: controller_interface.DelegateInterface] = ref object of controller_interface.AccessInterface
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/controller.nim:13]: yeah i got it. Thanks!

File: src/app/modules/main/node_section/module.nim [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import ../../../../../app_service/service/network/service as network_service
+import ../../../../../app_service/service/settings/service as settings_service
+
+import eventemitter
+
+export io_interface
+
+type
+  Module* [T: io_interface.DelegateInterface] = ref object of io_interface.AccessInterface
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/module.nim:14]: got it.

File: src/app/modules/main/node_section/module.nim [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  settingService: setting_service.Service):
+  Module[T] =
+  result = Module[T]()
+  result.delegate = delegate
+  result.view = view.newView(result)
+  result.viewVariant = newQVariant(result.view)
+  result.controller = controller.newController[Module[T]](result, events, contactsService, accountsService)
+  result.moduleLoaded = false
+
+method delete*(self: Module) =
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/module.nim:33]: okay got it.

File: src/app/modules/main/node_section/module.nim [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  singletonInstance.engine.setRootContextProperty("nodeSection", self.viewVariant)
+  self.view.load()
+
+method isLoaded*(self: Module): bool =
+  return self.moduleLoaded
+
+method viewDidLoad*(self: Module) = 
+  self.moduleLoaded = true
+
+## iview interfaces
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/module.nim:47]: okay :)

**Comment by borismelnik** [comment]
> @borismelnik I added some observations, we may discuss them further if needed.
> 
> * The main thing you were able to build the project and why you haven't seen errors is cause your module has not been created/called ever. So bring it to the app and connect it appropriately.
> * One more general suggestion is that setting delegate for a child module, in case of Nim, is maybe easier/cleaner using `AccessInterface` than the concepts, and also that will enable using third, fourth levels of submodules if needed, while with concepts only 2 sub levels is doable.
> * Another general thing is about naming interfaces files in `private_interfaces`, I don't have anything against the way you used, but maybe we should stick to the naming pattern already used in other `private_interfaces` folders.

1. Yes i never use it bc its new for me and i provide this PR just for understanding that i'm on the right way.
2. Yes, agreed :)
3. Sorry about it, but i was thinking that yours `new way`? Maybe i'm not get it?

File: src/app/modules/main/node_section/controller.nim [hunk 5]
```diff
@@ -0,0 +1,84 @@
+import ./controller_interface
+import io_interface
+import ../../../../../app_service/service/network/service as network_service
+import ../../../../../app_service/service/settings/service as settings_service
+import ../../../../../app_service/service/node/service as node_service
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/controller.nim:5]: Should i create this service?

File: src/app_service/service/node/service.nim [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    proc bitsSet*(self: Service, bitsSet: string) {.slot.} =
+        self.bloomBitsSet = parseInt(bitsSet)
+        self.events.emit(SIGNAL_BITS_SET_FETCHED, BitsSet(bitsSet: self.bloomBitsSet))
+
+    proc sendRPCMessageRaw*(self: Service, msg: string): string =
+        echo "sending RPC message"
+        status_core.callPrivateRPCRaw(msg)
+
+    proc adminPeers*(): seq[string] =
+        let response = status_core.callPrivateRPC("admin_peers", %* []).result
```
Comment (saledjenic) [inline src/app_service/service/node/service.nim:69]: @borismelnik calls like this should go to `status-lib`, so move them there. Also take a look how it's done in other services.

File: src/app_service/service/node/service.nim [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        for jsonPeer in response:
+            result.add(jsonPeer["enode"].getStr)
+
+    proc wakuV2Peers*(): seq[string] =
+        let response = status_core.callPrivateRPC("peers".prefix, %* []).result
```
Comment (saledjenic) [inline src/app_service/service/node/service.nim:74]: @borismelnik same as above.

File: src/app_service/service/node/service.nim [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    proc getBloomBitsSet*(self: Service): int =
+        return self.bloomBitsSet;
+
+    proc bitsSet*(self: Service, bitsSet: string) {.slot.} =
+        self.bloomBitsSet = parseInt(bitsSet)
+        self.events.emit(SIGNAL_BITS_SET_FETCHED, BitsSet(bitsSet: self.bloomBitsSet))
+
+    proc sendRPCMessageRaw*(self: Service, msg: string): string =
+        echo "sending RPC message"
+        status_core.callPrivateRPCRaw(msg)
```
Comment (saledjenic) [inline src/app_service/service/node/service.nim:66]: @borismelnik is this what you wanted? That method shouldn't be called from here directly, it's private for `status-lib`.

File: src/app/modules/main/node_section/controller_interface.nim [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+method delete*(self: AccessInterface) {.base.} =
+  raise newException(ValueError, "No implementation available")
+
+method init*(self: AccessInterface) {.base.} =
+  raise newException(ValueError, "No implementation available")
+
+method sendRPCMessageRaw*(self: AccessInterface, inputJSON: string): string {.base.} =
+  raise newException(ValueError, "No implementation available")
+
+method setBloomFilterMode*(self: AccessInterface, bloomFilterMode: bool): StatusGoError {.base.} =
```
Comment (saledjenic) [inline src/app/modules/main/node_section/controller_interface.nim:17]: @borismelnik `RpcResponse` doesn't contain `StatusGoError` remove it should not be used in this context. State for all places.

File: src/app/modules/main/node_section/controller_interface.nim [hunk 10]
```diff
@@ -0,0 +1,36 @@
+import json
+import status/types/[rpc_response]
```
Comment (saledjenic) [inline src/app/modules/main/node_section/controller_interface.nim:2]: @borismelnik new refactored part should not rely on the old code from `status-lib`.

**Review by saledjenic — CHANGES_REQUESTED** [review]

File: src/app/modules/main/node_section/controller_interface.nim (hunk 10 above)
Comment (borismelnik) [reply src/app/modules/main/node_section/controller_interface.nim:2]: got it

File: src/app/modules/main/node_section/controller_interface.nim (hunk 9 above)
Comment (borismelnik) [reply src/app/modules/main/node_section/controller_interface.nim:17]: got it

File: src/app_service/service/node/service.nim [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+            self.connected = true
+            self.events.emit("network:connected", Args())
+
+        self.peers = peers
+
+    proc peerCount*(self: Service): int = self.peers.len
+
+    proc addNetwork*(self: Service, name: string, endpoint: string, networkId: int, networkType: string) =
+        discard
+        # var networks = self.settingsService.getAvailableNetworks()
```
Comment (saledjenic) [inline src/app_service/service/node/service.nim:105]: @borismelnik what's with this?

File: src/app/modules/main/node_section/view.nim [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+
+  QtProperty[string] bloomLevel:
+    read = getBloomLevel
+    notify = initialized
+
+  proc setWakuBloomFilterMode*(self: View, bloomFilterMode: bool) {.slot.} =
+    self.delegate.setBloomFilterMode(bloomFilterMode)
+    quit(QuitSuccess) # quits the app TODO: change this to logout instead when supported
+
+  proc init*(self: View) {.slot.} =
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:97]: @borismelnik not sure that I understand purpose of this proc and where is it called from?

File: src/app/modules/main/node_section/view.nim [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+    self.wakuBloomFilterMode = self.delegate.getWakuBloomFilterMode()
+    self.fullNode = self.delegate.isFullNode()
+    self.wakuV2LightClient = self.delegate.isV2LightMode()
+    self.initialized()
+
+  proc wakuVersion*(self: View): int {.slot.} =
+    var fleetStr = self.delegate.getFleet()
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:104]: @borismelnik this is the logic for module/controller, move it there.

File: src/app/modules/main/node_section/view.nim [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+      BloomFilterMode = false
+      FullNode = false
+    of "full":
+      BloomFilterMode = true
+      FullNode = true
+    else:
+      BloomFilterMode = true
+      FullNode = false
+
+    self.delegate.setBloomLevel(BloomFilterMode, FullNode)
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:124]: @borismelnik just call `self.delegate.setBloomLevel(BloomFilterMode, FullNode)` and apply logic in module/controller.

File: src/app/modules/main/node_section/view.nim [hunk 15]
```diff
@@ -0,0 +1,188 @@
...
+  proc getBloomBitsSet(self: View): int {.slot.} =
+    self.bloomBitsSet
+
+  proc bloomBitsSetChanged(self: View) {.signal.}
+
+  proc setBitsSet*(self: View, bitsSet: int) =
+    self.bloomBitsSet = bitsSet
+    self.bloomBitsSetChanged();
+
+  QtProperty[int] bloomBits:
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:146]: @borismelnik this `bloomBits` and corresponding methods are perfect, exactly as it should be.

File: src/app/modules/main/node_section/module.nim [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+method load*(self: Module) =
+  singletonInstance.engine.setRootContextProperty("nodeModel", self.viewVariant)
+  self.controller.init()
+  self.view.load()
+
+method isLoaded*(self: Module): bool =
+  return self.moduleLoaded
+
+method viewDidLoad*(self: Module) =
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/module.nim:53]: got it

File: src/app/modules/main/node_section/view.nim [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+  proc `callResult=`*(self: View, callResult: string) = self.setCallResult(callResult)
+
+  QtProperty[string] callResult:
+    read = callResult
+    write = setCallResult
+    notify = callResultChanged
+
+  proc onSend*(self: View, inputJSON: string) {.slot.} =
+    self.setCallResult(self.delegate.sendRPCMessageRaw(inputJSON))
+    echo "Done!: ", self.callResult
```
[the remark this answers is not in the document]
Comment (borismelnik) [reply src/app/modules/main/node_section/view.nim:56]: @saledjenic are you mean `echo`?

File: src/app_service/service/node/service.nim (hunk 11 above)
Comment (borismelnik) [reply src/app_service/service/node/service.nim:105]: @saledjenic oh yes i forgot... do you know how i can add new network now? I did't find any suitable methods.

File: src/app/modules/main/node_section/view.nim (hunk 13 above)
Comment (borismelnik) [reply src/app/modules/main/node_section/view.nim:104]: got it

File: src/app/modules/main/module.nim [hunk 18]
```diff
@@ -340,6 +350,9 @@ proc checkIfModuleDidLoad [T](self: Module[T]) =
   if(not self.appSearchModule.isLoaded()):
     return
 
+  if(not self.nodeSectionModule.isLoaded()):
+    return
```
Comment (jrainville) [inline src/app/modules/main/module.nim:354]: This is duplicated

File: src/app_service/service/node/async_tasks.nim [hunk 19]
```diff
@@ -0,0 +1,14 @@
+import status/status
```
Comment (jrainville) [inline src/app_service/service/node/async_tasks.nim:1]: Sale already pointed it out, but we shouldn't rely on the old `status-lib` code. IMO, we can do it in another PR. Just need to open an issue to keep it in mind.

File: src/app_service/service/node/service.nim [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+            return adminPeers()
+
+    proc peerSummaryChange*(self: Service, peers: seq[string]) =
+        if peers.len == 0 and self.connected:
+            self.connected = false
+            self.events.emit("network:disconnected", Args())
+        
+        if peers.len > 0 and not self.connected:
+            self.connected = true
+            self.events.emit("network:connected", Args())
```
Comment (jrainville) [inline src/app_service/service/node/service.nim:96]: We can use `const`s for those event names

**Review by jrainville — COMMENTED** [review]
Most of it looks good. 

I think the use of old status-lib code should be removed, but it's ok to do it another time.

File: src/app/modules/main/node_section/view.nim [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+    result = self.callResult
+
+  proc callResultChanged*(self: View, callResult: string) {.signal.}
+
+  proc setCallResult(self: View, callResult: string) {.slot.} =
+    if self.callResult == callResult: return
+    self.callResult = callResult
+    self.callResultChanged(callResult)
+
+  proc `callResult=`*(self: View, callResult: string) = self.setCallResult(callResult)
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:39]: @borismelnik what's the point of this proc? What's its usage (currently I don't see it, if you don't need it remove it)?

File: src/app/modules/main/node_section/view.nim [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+
+  proc load*(self: View) =
+    self.delegate.viewDidLoad()
+
+  proc callResult*(self: View): string {.slot.} =
+    result = self.callResult
+
+  proc callResultChanged*(self: View, callResult: string) {.signal.}
+
+  proc setCallResult(self: View, callResult: string) {.slot.} =
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:34]: @borismelnik this don't need to be a slot.

File: src/app/modules/main/node_section/view.nim [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+    if self.callResult == callResult: return
+    self.callResult = callResult
+    self.callResultChanged(callResult)
+
+  proc `callResult=`*(self: View, callResult: string) = self.setCallResult(callResult)
+
+  QtProperty[string] callResult:
+    read = callResult
+    write = setCallResult
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:43]: @borismelnik remove `write` prop, as it's `callResult` is used only to be read on the qml side.

File: src/app_service/service/node/async_tasks.nim [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+include ../../../app/core/tasks/common
+
+type
+  BloomBitsSetTaskArg = ref object of QObjectTaskArg
+    bitsSet: int
+
+const bloomBitsSetTask: Task = proc(argEncoded: string) {.gcsafe, nimcall.} =
+  let
+    arg = decode[BloomBitsSetTaskArg](argEncoded)
+    output = getBloomFilterBitsSet(nil)
```
Comment (saledjenic) [inline src/app_service/service/node/async_tasks.nim:12]: @borismelnik this is not moved to new `status-lib` section.

File: src/app/modules/main/node_section/view.nim [hunk 25]
```diff
@@ -0,0 +1,150 @@
...
+
+  proc onSend*(self: View, inputJSON: string) {.slot.} =
+    self.setCallResult(self.delegate.sendRPCMessageRaw(inputJSON))
+
+  proc onMessage*(self: View, message: string) {.slot.} =
+    self.setCallResult(message)
+
+  proc receivedMessage*(self: View, lastMessage: string) {.signal.}
+
+  proc setLastMessage*(self: View, lastMessage: string) {.slot.} =
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:54]: @borismelnik this doesn't need to be a slot.

File: src/app/modules/main/node_section/view.nim [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+    self.lastMessage = lastMessage
+    self.receivedMessage(lastMessage)
+
+  QtProperty[string] lastMessage:
+    read = lastMessage
+    write = setLastMessage
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:60]: @borismelnik remove `write` property.

File: src/app/modules/main/node_section/view.nim [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+    notify = receivedMessage
+
+  proc getWakuBloomFilterMode*(self: View): bool {.slot.} =
+    result = self.delegate.getWakuBloomFilterMode()
+
+  proc getBloomLevel*(self: View): string {.slot.} =
+    var wakuBloomFilterMode = self.delegate.getWakuBloomFilterMode()
```
Comment (saledjenic) [inline src/app/modules/main/node_section/view.nim:67]: @borismelnik this logic is not for the view, actually there is no any more `getWakuBloomFilterMode` method, seems this is not rebased for a while, now we have `getBloomLevel` containing all logic there, you just need to call it.

**Review by saledjenic — CHANGES_REQUESTED** [review]
@borismelnik as a first step rebase this onto latest `base_bc` and updated code accordingly.
Then update the code following @jrainville and my comments here and re-request a review once all that is done.

**Review by saledjenic — APPROVED** [review]

File: src/app/modules/main/node_section/controller.nim [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+    self.setPeers(data.peers)
+
+  self.events.on(SignalType.Stats.event) do (e:Args):
+    self.delegate.setStats(StatsSignal(e).stats)
+    if not self.isWakuV2: self.delegate.fetchBitsSet()
+
+  self.events.on(SignalType.ChroniclesLogs.event) do(e:Args):
+    self.delegate.log(ChroniclesLogsSignal(e).content)
+
+  self.events.on(SIGNAL_BITS_SET_FETCHED) do (e:Args):
```
Comment (richard-ramos) [inline src/app/modules/main/node_section/controller.nim:65]:
```suggestion
  self.events.on(SIGNAL_BLOOMFILTER_BITS_SET_FETCHED) do (e:Args):
```

File: src/app_service/service/node/service.nim [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+import ../../../app/core/tasks/[qt, threadpool]
+import ../../../app/core/fleets/fleet_configuration
+
+include async_tasks
+
+logScope:
+  topics = "node-service"
+
+# Signals which may be emitted by this service:
+const SIGNAL_BITS_SET_FETCHED* = "SIGNAL_BITS_SET_FETCHED"
```
Comment (richard-ramos) [inline src/app_service/service/node/service.nim:17]:
```suggestion
const SIGNAL_BLOOMFILTER_BITS_SET_FETCHED* = "SIGNAL_BLOOMFILTER_BITS_SET_FETCHED" 
```

**Review by richard-ramos — APPROVED** [review]
